### PR TITLE
Add torch_lazy_all_numbers_special_scalars flag

### PR DESCRIPTION
**Pull request by antoniojkim**
This is to allow even non zero and one scalars to appear as constants in the graph. The assumption being that none of them will change.

The flag is set to `false` by default to preserve the original behaviour.

CC: @wconstab @JackCaoG @ke1337 @vaibhavc-cerebras @glebk-cerebras

### CODE REVIEW

**Comment by wconstab** [comment]
I think this change is probably OK insofar as it looks harmless.

Is the intent that all scalars would be burned into the graph rather than passed into the compiled function?

I would expect this makes even minor forms of dynamism such as adam optimizer learning rate stuff into a recompile for you.  Maybe you're not using the standard optimizers and have another plan here.

any concerns @JackCaoG @Krovatkin?

**Comment by antoniojkim** [comment]
> Is the intent that all scalars would be burned into the graph rather than passed into the compiled function?

Yes, thats correct.

> I would expect this makes even minor forms of dynamism such as adam optimizer learning rate stuff into a recompile for you. Maybe you're not using the standard optimizers and have another plan here.

Thats correct. Our backend does not yet support passing in constants as inputs so we have custom optimizers and learning rate schedulers to account for this.

The plan is to eventually support passing constants as inputs eventually, so hopefully we can get away from this at some point in the future

File: torch/csrc/lazy/core/tensor_util.cpp [hunk 1]
```diff
@@ -60,6 +60,9 @@ std::vector<BackendDataPtr> CreateTensorsData(
 bool IsSpecialScalar(const at::Scalar& value) {
   if (FLAGS_torch_lazy_handle_special_scalars &&
```
Comment (vaibhavc-cerebras) [inline torch/csrc/lazy/core/tensor_util.cpp:61]: should this be:
if ((FLAGS_torch_lazy_handle_special_scalars || FLAGS_torch_lazy_all_numbers_special_scalars) &&

Comment (antoniojkim) [reply torch/csrc/lazy/core/tensor_util.cpp:61]: They can just both be set to `true`

**Comment by alanwaketan** [comment]
This change doesn't seem to be useful for the XLA backend. If it's only applicable for a specific backend for a specific period, I guess it's better to keep it away from the upstream? Thoughts?

**Comment by antoniojkim** [comment]
> This change doesn't seem to be useful for the XLA backend. If it's only applicable for a specific backend for a specific period, I guess it's better to keep it away from the upstream? Thoughts?

@alanwaketan Where else could it be put? We want to be able to use the release version of PyTorch and not a local fork.

The default behaviour is preserved as the flag is `false` by default. So, it won't affect the XLA backend in any way. Also, XLA isn't the only backend that will use LTC. So I think there's a case to be made for having features in upstream that will allow other backends to be supported.

**Comment by alanwaketan** [comment]
What if somehow our customers have misconfigured their environment to have this option turned on, which could then create a hard to debug issue for us.

**Comment by alanwaketan** [comment]
Instead of having a generic runtime flag, something to consider is to pipe this setting through your own API maybe?

**Comment by antoniojkim** [comment]
> What if somehow our customers have misconfigured their environment to have this option turned on, which could then create a hard to debug issue for us.

@alanwaketan This could be said about any the gflags in this or any other config file. There's no way to completely avoid that possibility. Having said that, I think its unlikely that a customer would know about this flag to begin with since its an implementation detail on the C++ side. So, it should be fairly safe.

> Instead of having a generic runtime flag, something to consider is to pipe this setting through your own API maybe?
>
> Another way is to add this flag to one of the backend API? Such that each backend can configure it separately and not user facing?

We could do that, but I think that solution is overkill. For something that is a simple control flow boolean, making a completely new backend interface method seems unnecessarily complicated.  CC: @wconstab

**Comment by alanwaketan** [comment]
I agreed that it's a mess already. But that doesn't mean we should make it even more messy?

Generally speaking, I think we should be more conservative on introducing new runtime flags. They are hard to use and maintain. Two things I would consider to be suitable as runtime flags:
1. experimental features that change a lot of common code paths and which we will eventually turn on by default.
2. developer facing debug utils.

**Comment by antoniojkim** [comment]
> I agreed that it's a mess already. But that doesn't mean we should make it even more messy?

Perhaps not. But I think this discussion is outside the scope of the PR. Should this be cleaned up at some in the future, I think it can all be addressed at once. Adding this flag wouldn't make that process any more difficult than it already is.
 
> Generally speaking, I think we should be more conservative on introducing new runtime flags. They are hard to use and maintain. Two things I would consider to be suitable as runtime flags:
> 
>     1. experimental features that change a lot of common code paths and which we will eventually turn on by default.
> 
>     2. developer facing debug utils.

I think another category to consider here is legacy code that will eventually be deprecated. The flag would keep that code path "turned off" by default but still remains for backwards compatibility.

The flag being introduced in this PR would fit into that category. Eventually, we do plan on being able to handle scalars as inputs. But we don't have that capability at this time. Thus, for backwards compatibility, we would like to be able to treat all scalars specially as constants.

**Comment by alanwaketan** [comment]
Well, basically you are trying to introduce some hacks that workaround your backend's limitations to upstream, right?

**Comment by antoniojkim** [comment]
> Well, basically you are trying to introduce some hacks that workaround your backend's limitations to upstream, right?

I would not call it a hack. It is no more hacky than the currently existing [torch_lazy_handle_special_scalars](https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/core/config.h#L6). The idea of having a flag to control the flow of tracing is nothing new. Its just a new case to handle a limitation in our backend.

**Comment by alanwaketan** [comment]
Yea, we should come out a way to improve the current situation.

**Comment by antoniojkim** [comment]
> Yea, we should come out a way to improve the current situation.

I don't disagree. But that is way beyond the scope of this PR.

**Review by wconstab — APPROVED** [review]
I think it is reasonable to enable the C backend to handle scalars however they want, and the only 2 ways to do that are an ENV or a new backend API.  given backend API change would be more visible and impactful i think the ENV way is better.

**Comment by alanwaketan** [comment]
I think the feature request is valid, and I'm just not a big fan of using ENV.
